### ISSUE TITLE
Connection errors/qcache errors after upgrading from 0.5.2 to 0.5.3

### ISSUE THREAD

**Issue by rdunlop** [brianmario/mysql2#1090]
I upgraded to 0.5.3 earlier this week from 0.5.2, and upon deployment to production, my app went down.

Luckily, reverting to the previous release brought things back up, though the mysql query cache was screwed up. (I am not sure if this is a related problem or not).

Here's the original problem that I'm seeing when trying to load the main page:
```
F, [2019-12-04T20:20:37.247797 #15929] FATAL -- : [53a2b69a-1981-4883-852f-cfd2527a3409]   
F, [2019-12-04T20:20:37.247846 #15929] FATAL -- : [53a2b69a-1981-4883-852f-cfd2527a3409] ActionView::Template::Error (Mysql2::Error::ConnectionError: Lost connection to MySQL server during query: SHOW FULL FIELDS FROM `academic_years`):
```

When using the rails console, I see some other strange behaviors:
```
> School.count

ActiveRecord::StatementInvalid (Mysql2::Error::ConnectionError: Lost connection to MySQL server during query: SHOW FULL FIELDS FROM `schools`)
```

and also
```
> AcademicYear.count
Traceback (most recent call last):
        1: from (irb):1
ActiveRecord::StatementInvalid (Mysql2::Error: : SELECT COUNT(*) FROM `academic_years`
```

...strangely, I was able to resolve the above by performing `RESET QUERY CACHE` on the mysql server, but upon upgrading to 0.5.3 again, I continue to have problems.

I'm running:
- Rails 5.2.3
- Ruby 2.5.0
- mysql2 0.5.3
- AWS RDS MySql: Server version: 5.7.19-log MySQL Community Server
- Amazon Linux AMI 2018.03 based server image

If you need more details about my exact configuration, please let me know.

Further frustrating, this works correctly on my staging server, which is nearly identical to the prod server (same, rails, same ruby, same mysql2, same AMI, but mysql server is 5.7.17, that can't be it, can it?).

My current theory is that it may be related to the ruby 2.5.0, as I have another application running 2.5.1 (a different application, on a different database on the same db instance), and it doesn't seem to have the same problem. I have not had a chance to test this theory yet, as upgrading ruby versions in prod (since it only happens in prod) is not something I can do easily.

**Comment by rdunlop** [comment]
Update: I was mistaken, my other application which also connects to this same mysql server, has the same problem when upgraded to 0.5.3. The version of ruby doesn't appear to matter.

I've checked the mysql parameter group for prod and staging, and they are identical. I  don't know what to do next, so far, I've been forced to lock to 0.5.2.

**Comment by ni1337** [comment]
I've ran in a similar problem. I've upgraded from 0.5.2 to 0.5.3 and got different exceptions:

```
Mysql2::Error::ConnectionError: Lost connection to MySQL server during query
Mysql2::Error::ConnectionError: Malformed packet
```

It seems the MySQL connection can not be created.

BUT: This works with MariaDB version 10.1.29, but not with 10.1.37.

**Comment by rdunlop** [comment]
I've run across this again, same machine....but I didn't upgrade the `mysql` gem this time. This time I was upgrading the `ruby` version on the machine. After upgrading ruby, and `bundle`-ing again, this issue has appeared again.

For this reason, I suspect that the issue is not related to 0.5.3, but is instead related to something else which must have changed on the machine since the time when 0.5.2 was installed. 

Perhaps this is related to https://github.com/brianmario/mysql2/issues/938#issuecomment-425953296, and so I'm looking at my libraries... I see:
```
 Dec  3  2018 libmysqlclient.so.18.0.0
```

Since mysql2 0.5.2 was released July 2018, ...if the libmysqlclient.so.18.0.0 was updated AFTER that, it could make sense that 0.5.3 would appear to have issues, because it's trying to compile against a different libmysqlclient library....one which 0.5.2 was not compiled against....if that is somehow a problem.

I will try to continue noting things here as I learn more.

**Comment by rdunlop** [comment]
So, both staging and production nodes use the same libmysqlclient library version (not the one listed in my previous comment)....So that's not it.

**Comment by rdunlop** [comment]
I think that I'm at the end of my investigation, I have found that setting `strict: false` in my database.yml has resolved the issue. It appears that when mysql2 was recompiled (either due to a new version of ruby, or a new version of mysql2), it linked against a different version of libmysqlclient.

Now https://dzone.com/articles/upgrading-to-mysql-57-beware-of-the-new-strict-mod warns about strict mode...so....

long story short, setting `strict: false` has resolved my issue.
